Zend/zend_API.c

```diff
@@ -1173,7 +1168,22 @@ ZEND_API int zend_update_class_constants(zend_class_entry *class_type) /* {{{ */
 						val = (zval*)((char*)class_type->default_properties_table + prop_info->offset - OBJ_PROP_TO_OFFSET(0));
 					}
 					if (Z_TYPE_P(val) == IS_CONSTANT_AST) {
-						if (UNEXPECTED(zval_update_constant_ex(val, ce) != SUCCESS)) {
+						if (prop_info->type) {
+							zval tmp;
+
+							ZVAL_COPY(&tmp, val);
+							if (UNEXPECTED(zval_update_constant_ex(&tmp, ce) != SUCCESS)) {
+								return FAILURE;
+							}
+							if (UNEXPECTED(!zend_verify_property_type(prop_info, &tmp, ZEND_CALL_USES_STRICT_TYPES(EG(current_execute_data))))) {
```

Comment:
I'm unsure about this on. One one hand this is inconsistent with how a non-constant initializer is treated:

```php
class Test {
    public int $x = '42'; // illegal
    public int $x = FOO; // currently legal
}
define('FOO', '42');
var_dump(new Test);
```

On the other hand, this is consistent with how constants are treated in parameters:

```
function foo(int $x = FOO) { // currently legal
    var_dump($x);
}
define('FOO', '42');
foo();
```

Overall I think that the existing behavior for parameter types is a bug, because there will be inconsistent behavior depending on whether the constant can be resolved at compile time or not. We should always use strict types here to be consistent with the compile-time behavior.

Reply:
There's anyway a problem with this; it takes the strict_types setting of the active file context when the constant expression is resolved instead of the strict_types of the file containing the class.

I'm not sure on whether using strict types by default is the best way to go, at least with constants one does not have immediate control over the value, whereby specifying a literal is unambiguous.

If I got to choose I'd always use the strict_types setting of the declaring file for constants so that behavior will always be consistent independently of compile-time or run-time resolution.

Reply:
The main problem I see with not using strict_types always is that these two cases will behave differently:

```
class Test {
    const FOO = '42';
    public int $x = self::FOO; // illegal
}
class Test {
    public int $x = self::FOO; // legal
    const FOO = '42';
}
```

These *should* work the same way, but because one will be resolved at compile-time and the other at run-time the behavior ends up being inconsistent.

Reply:
As said, "always use the strict_types setting of the declaring file" - independently of whether it's resolved at compile-time or at run-time. The compile time resolution is an implementation detail here, which the user shouldn't have to know of.